doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,414 @@
+...     pass
+```
+
+At this point, `versin` has no behaviors defined on it. It is very similar to
+an [undefined function](custom-functions-fully-symbolic). We note that
+`versin` is a class, and `versin(x)` is an instance of this class.
+
+```
+>>> type(versin)
+<class 'sympy.core.function.FunctionClass'>
+>>> versin(x)
+versin(x)
+>>> isinstance(versin(x), versin)
+True
+```
+
+### Defining Automatic Evaluation with `eval`
+
+
+```{sidebar} Reminder
+Remember that `eval` should be defined with the `@classmethod` decorator.
+```
+
+The first and most common thing we might want to define on our custom function
+is to define automatic evaluation, that is, cases where it will return an
+actual value instead of just remaining unevaluated as-is.
+
+This is done by defining the class method `eval`. `eval` should take the
+arguments of the function and return either a value or `None`. If it returns
+`None`, the function will remain unevaluated in that case.
+
+For our function `versin`, we might recall that $\cos(n\pi) = (-1)^n$ for
+integer $n$, so $\operatorname{versin}(n\pi) = 1 - (-1)^n.$ We can make
+`versin` automatically evaluate to this value when passed an integer multiple
+of `pi`.
+
+```
+>>> from sympy import pi, Integer
+>>> class versin(Function):
+...    @classmethod
+...    def eval(cls, x):
+...        # If x is an integer multiple of pi, x/pi will cancel and be an Integer
+...        n = x/pi
+...        if isinstance(n, Integer):
+...            return 1 - (-1)**n
+>>> versin(pi)
+2
+>>> versin(2*pi)
+0
+>>> versin(x*pi)
+versin(pi*x)
+```
+
+Note that in this example, we make use of the fact that if a Python function
+does not explicitly return a value, it automatically returns `None`, so in the
+cases where the `if` statement is not triggered, `eval` returns `None` and
+`versin` remains unevaluated.
+
+`eval` can take any number of arguments, including an arbitrary number with
+`*args` and optional keyword arguments. The `.args` of the function will
+always be the arguments that were passed in by the user. For example
+
+```py
+>>> class f(Function):
+...     @classmethod
+...     def eval(cls, x, y=1, *args):
+...         return None
+>>> f(1).args
+(1,)
+>>> f(1, 2).args
+(1, 2)
+>>> f(1, 2, 3).args
+(1, 2, 3)
+```
+
+#### Best Practices for `eval`
+
+- **Don't just return an expression.**
+
+  We might have been tempted to write
+
+  ```py
+  >>> from sympy import cos
+  >>> class versin(Function):
+  ...    @classmethod
+  ...    def eval(cls, x):
+  ...        return 1 - cos(x)
+  ```
+
+  However, this would make it so that `versin(x)` would *always* return `1 -
+  cos(x)`, regardless of what `x` is. If all you want is a quick shorthand to
+  `1 - cos(x)`, that is fine. But would be much simpler and more explicit to
+  just use a Python function, as [described
+  above](custom-functions-fully-evaluated). If we defined `versin` like this,
+  it would never actually return as `versin(x)`, and none of the other
```

Comment:
Unless `evaluate=False` is given. Perhaps worth mentioning here and also to clarify that `__new__` should not be used for `Function` subclasses because `Function.__new__` already takes care of things like this.

Reply:
Good point, I'll mention to avoid `__init__` and `__new__`. 

Actually it would be nice if the `eval` logic were on `Basic` rather than `Function`. It might have to be generalized a bit, but it would be great it 99% of SymPy objects did not need to define `__new__`.